neuralpredictors/training/tracking.py

```diff
@@ -146,3 +155,372 @@ def asdict(self, time_key="time", make_copy=True):
+     - Scores can be automatically normalized if `"normalization"` (e.g. total examples) is tracked on the same hierarchy level
+     - Automatic per-epoch initialization for each objective when a new epoch is started via `start_epoch()`
+     - Automatic output of current logs to a logger or even tqdm
+     - Easy extraction and reloading of tracker state via `state_dict()`, `from_dict(sd)` and `load_state_dict(sd)`
+
+
+     Example usage:
+     1. Specify the log structure and initial values:
+         ```
+        objectives = {
+            "lr": 0,
+            "training": {
+                "img_classification": {"loss": 0, "accuracy": 0, "normalization": 0}
+            },
+            "validation": {
+                "img_classification": {
+                    "loss": 0,
+                    "accuracy": 0,
+                    "normalization": 0,
+                },
+                "patience": 0,
+            },
+        }
+        tracker = AdvancedTracker(
+            main_objective=("img_classification", "accuracy"), **objectives
+        )
+         ```
+     2. Start a new epoch:
+     ```
+        self.tracker.start_epoch()
+     ```
+     3. Log the objectives:
+     ```
+        tracker.log_objective(
+                100 * predicted.eq(targets).sum().item(),
+                keys=(mode, task_key, "accuracy"),
+            )
+        tracker.log_objective(
+            batch_size,
+            keys=(mode, task_key, "normalization"),
+        )
+        tracker.log_objective(
+            loss.item() * batch_size,
+            keys=(mode, task_key, "loss"),
+        )
+    ```
+    4. Display the current objective values, e.g. everything related to training:
+    ```
+        self.tracker.display_log(tqdm_iterator=t, keys=("training",))
+    ```
+    5. Save the tracker (i.e. save the training progress):
+    ```
+        self.tracker.state_dict()
+    ```
+    """
+
+    def __init__(self, main_objective: Tuple[str, ...] = (), **objectives: Mapping) -> None:
+        """
+        In principle, `objectives` is expected to be a dictionary of dictionaries
+        The hierarchy can in principle be arbitrarily deep.
+        The only restriction is that the lowest level has to be a dictionary with values being
+        either a numerical value which will be interpreted as the initial value for this objective
+        (to be accumulated manually) or a callable (e.g. a function) that returns the objective value.
+
+        Args:
+            main_objective: key of the main objective that is used to e.g. decide lr reductions
+                            can be something like `("img_classification","accuracy")` to always look at accuracy.
+                            Can be combined with a setting specific key in `get_current_main_objective(...)`
+            **objectives: e.g. {"dataset": {"objective1": o_fct1, "objective2": 0, "normalization": 0},...}
```

Comment:
Could you extend the docstring a bit to elaborate on the `normalization` I have troubles understanding what it does and how it is used.

Reply:
good point